// youtube-comments-task/dist/lib/fetch-replies.js
'use strict';

var Task = require('data.task');
var Either = require('data.either');
var debug = require('debug')('fetch-replies');

var _require = require('./youtube-api/youtube-api'),
    commentReplies = _require.commentReplies;

var _require2 = require('./utils/cheerio-utils'),
    cheerio = _require2.cheerio;

var parseReplies = require('./parse-replies');

var _require3 = require('./error-handler'),
    scraperError = _require3.scraperError;

var getRepliesToken = function getRepliesToken(comment) {
  return Either.fromNullable(comment.repliesToken).leftMap(function (e) {
    return 'Comment parameter object does not have a repliesToken field';
  }).fold(Task.rejected, Task.of);
};

var getContentHtml = function getContentHtml(r) {
  return Either.fromNullable(r.content_html).leftMap(function (_) {
    return 'Invalid Replies-API response, does not contain content_html field';
  }).fold(Task.rejected, Task.of);
};

var parseCommentReplies = function parseCommentReplies($replies) {
  return parseReplies($replies).orElse(function (e) {
    debug('Unable to parse comment replies: %s', e);
    return Either.of([]);
  }).fold(Task.rejected, Task.of);
};

var fetchReplies = function fetchReplies(videoId, comment) {
  return getRepliesToken(comment).chain(function (repliesToken) {
    return commentReplies(videoId, repliesToken);
  }).chain(getContentHtml).map(function (html) {
    return cheerio('<div>' + html + '</div>');
  }).chain(parseCommentReplies).rejectedMap(function (e) {
    return e.type ? e : scraperError({
      videoId: videoId,
      message: e,
      component: 'fetch-replies',
      operation: 'fetch-replies'
    });
  });
};

module.exports = fetchReplies;
